### ISSUE TITLE
ComfyUI-Zluda corrupting itself randomly and causing python crash at startup

### ISSUE THREAD

**Issue by ReiskaC** [patientx/ComfyUI-Zluda#41]
### Your question

Running on Windows 11 Pro, git pulled & install.bat installed, Radeon 7900 XTX. Python 3.11.9 / HIP 5.7.1 / Adrenalin 24.10.1. ComfyUI-Zluda worked in the morning today (15.11.2024) and started crashing after I shut it down and tried to restart in the afternoon. git pull gives me 'Already up to date.' at this time. 
 
<img width="866" alt="2024-11-15 ComfyUI-Zluda crashing again on start-up" src="https://github.com/user-attachments/assets/f0c65224-c8f8-4c75-9343-e6586a673452">
 
My ComfyUI-Zluda somehow corrupts itself every other time I start the server and to regain operations is a totally undeterministic process (for me at least). 

Sometimes even when the comfyui.bat crashes at python I still can run it once by directly invoking 'zluda -- .\venv\Scripts\python.exe main.py', sometimes not. 

Usually on second restart after that won't work and crashes at start as well. 

I've tried  to use the 'fix-update.bat' 'patchzluda.bat' which none seem to fix this issue. 

I've also wiped the entire 'venv'-folder and reran 'install.bat' which sometimes resolves these issues. 

Other times I've had to go all the way back to start and uninstall everything from my computer (git, python, AMD drivers, etc.) and do a clean re-install from scratch before this resolves itself. 

Anyone have good ideas on how to resolve this in a deterministic way?

**Comment by patientx** [comment]
virus ? or hdd corruption ? only those two things come to mind.

**Comment by ReiskaC** [comment]
1. Performed a disk scan with Seatools -> passed, no SMART error. Less than 1 year old Nvme drive
2. Performed a full AV-scan with F-secure (which has been running on this machine since day 1) -> clean system
3. Tried starting Comfy from Command prompt instead of PowerShell - worked once and begun exhibiting same behavior after installed a custom node in Manager and restarted server (doesn't actually restart, just quits vs. ComfyUI main branch actually bounces the server from the GUI Restart-button.)
4. Now trying a full clean install of ComfyUI-Zluda from Command-prompt... lets see if this makes any difference.

Errors during startup after clean install:
### Loading: ComfyUI-Impact-Pack (V7.11.3)
[Impact Pack] Failed to import due to several dependencies are missing!!!!
Traceback (most recent call last):
  File "F:\StableDiffusion\ComfyUI-Zluda\nodes.py", line 2022, in load_custom_node
    module_spec.loader.exec_module(module)
  File "<frozen importlib._bootstrap_external>", line 940, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "F:\StableDiffusion\ComfyUI-Zluda\custom_nodes\ComfyUI-Impact-Pack\__init__.py", line 53, in <module>
    raise e
  File "F:\StableDiffusion\ComfyUI-Zluda\custom_nodes\ComfyUI-Impact-Pack\__init__.py", line 31, in <module>
    import impact.subpack_nodes  # This import must be done before cv2.
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "F:\StableDiffusion\ComfyUI-Zluda\custom_nodes\ComfyUI-Impact-Pack\impact_subpack\impact\subpack_nodes.py", line 3, in <module>
    import impact.core as core
  File "F:\StableDiffusion\ComfyUI-Zluda\custom_nodes\ComfyUI-Impact-Pack\modules\impact\core.py", line 7, in <module>
    from segment_anything import SamPredictor
ModuleNotFoundError: No module named 'segment_anything'

Cannot import F:\StableDiffusion\ComfyUI-Zluda\custom_nodes\ComfyUI-Impact-Pack module for custom nodes: No module named 'segment_anything'

**Comment by ReiskaC** [comment]
Done some more testing:
1. After the initial run of the install.bat in Command Prompt ComfyUI-Zluda will start once regardless of the errors above and if provided a checkpoint will generate the default bottle.
2. Once you CTRL-C + Y + Enter out of the server and try to restart it with ComfyUI.bat it will crash
3. After this you can once start the server with zluda -- python main.py, generate images and when you restart/shutdown the server this will start crashing like the comfyui.bat does
4. This happens even if you just do a clean git clone + install.bat and exit immediately after the GUI has loaded in the web browser
5. The custom_nodes\ComfyUI-Impact-Pack\config.ini has security level set to normal, which prohibits Manager to try fixing the package -> could this be the reason the sub_pack doesn't get installed during the install.bat script? 
6. running python install-manual.py in the impact pack folder seems to download the missing bits, but doesn't fix the startup error

Is there someway to hike up the log level of the start-up command to see any details on what actually fails? Now I just get the blank console with "Already up to date" and then the pop-up about python.exe failing and nothing else.

**Comment by patientx** [comment]
In the comfyuizluda folder in cmd

Venv/scripts/activate
pip install segment_anything

**Comment by ReiskaC** [comment]
Installed that and started ComfyUI for the first time after a clean install (modified the install.bat to not run it immediately):

(venv) F:\StableDiffusion\ComfyUI-Zluda>comfyui.bat
*** Checking and updating to new version if possible
remote: Enumerating objects: 6, done.
remote: Counting objects: 100% (6/6), done.
remote: Compressing objects: 100% (3/3), done.
remote: Total 6 (delta 3), reused 4 (delta 3), pack-reused 0 (from 0)
Unpacking objects: 100% (6/6), 2.84 KiB | 126.00 KiB/s, done.
From https://github.com/patientx/ComfyUI-Zluda
   3790865..9759287  master     -> origin/master
Updating 3790865..9759287
Fast-forward
 comfy_extras/nodes_model_merging_model_specific.py | 29 ++++++++++++++++++++++
 1 file changed, 29 insertions(+)

[START] Security scan
[DONE] Security scan
## ComfyUI-Manager: installing dependencies. (GitPython)
## ComfyUI-Manager: installing dependencies done.
** ComfyUI startup time: 2024-11-15 21:43:03.290461
** Platform: Windows
** Python version: 3.11.9 (tags/v3.11.9:de54cf5, Apr  2 2024, 10:12:12) [MSC v.1938 64 bit (AMD64)]
** Python executable: F:\StableDiffusion\ComfyUI-Zluda\venv\Scripts\python.exe
** ComfyUI Path: F:\StableDiffusion\ComfyUI-Zluda
** Log path: F:\StableDiffusion\ComfyUI-Zluda\comfyui.log

Prestartup times for custom nodes:
   5.6 seconds: F:\StableDiffusion\ComfyUI-Zluda\custom_nodes\ComfyUI-Manager

Total VRAM 24560 MB, total RAM 65244 MB
pytorch version: 2.3.0+cu118
Set vram state to: NORMAL_VRAM

***----------------------ZLUDA--------------------------***
  ::  ZLUDA detected, disabling non-supported functions.
  ::  (cuDNN, flash_sdp, mem_efficient_sdp disabled)
***-----------------------------------------------------***
  ::  Device: cuda:0 AMD Radeon RX 7900 XTX [ZLUDA] : native

Using sub quadratic optimization for cross attention, if you have memory or speed issues try using: --use-split-cross-attention
****** User settings have been changed to be stored on the server instead of browser storage. ******
****** For multi-user setups add the --multi-user CLI argument to enable multiple user profiles. ******
[Prompt Server] web root: F:\StableDiffusion\ComfyUI-Zluda\web
### Loading: ComfyUI-Impact-Pack (V7.11.3)
[Impact Pack] Failed to import due to several dependencies are missing!!!!
Traceback (most recent call last):
  File "F:\StableDiffusion\ComfyUI-Zluda\nodes.py", line 2022, in load_custom_node
    module_spec.loader.exec_module(module)
  File "<frozen importlib._bootstrap_external>", line 940, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "F:\StableDiffusion\ComfyUI-Zluda\custom_nodes\ComfyUI-Impact-Pack\__init__.py", line 53, in <module>
    raise e
  File "F:\StableDiffusion\ComfyUI-Zluda\custom_nodes\ComfyUI-Impact-Pack\__init__.py", line 31, in <module>
    import impact.subpack_nodes  # This import must be done before cv2.
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "F:\StableDiffusion\ComfyUI-Zluda\custom_nodes\ComfyUI-Impact-Pack\impact_subpack\impact\subpack_nodes.py", line 3, in <module>
    import impact.core as core
  File "F:\StableDiffusion\ComfyUI-Zluda\custom_nodes\ComfyUI-Impact-Pack\modules\impact\core.py", line 10, in <module>
    from impact.utils import *
  File "F:\StableDiffusion\ComfyUI-Zluda\custom_nodes\ComfyUI-Impact-Pack\modules\impact\utils.py", line 3, in <module>
    import cv2
ModuleNotFoundError: No module named 'cv2'

Cannot import F:\StableDiffusion\ComfyUI-Zluda\custom_nodes\ComfyUI-Impact-Pack module for custom nodes: No module named 'cv2'
### Loading: ComfyUI-Manager (V2.51.9)
### ComfyUI Revision: 3033 [9759287c] | Released on '2024-11-15'

Import times for custom nodes:
   0.0 seconds: F:\StableDiffusion\ComfyUI-Zluda\custom_nodes\ComfyUI-deepcache
   0.0 seconds: F:\StableDiffusion\ComfyUI-Zluda\custom_nodes\websocket_image_save.py
   0.0 seconds (IMPORT FAILED): F:\StableDiffusion\ComfyUI-Zluda\custom_nodes\ComfyUI-Impact-Pack
   0.6 seconds: F:\StableDiffusion\ComfyUI-Zluda\custom_nodes\ComfyUI-Manager

Starting server

To see the GUI go to: http://127.0.0.1:8188
[ComfyUI-Manager] default cache updated: https://raw.githubusercontent.com/ltdrdata/ComfyUI-Manager/main/alter-list.json
[ComfyUI-Manager] default cache updated: https://raw.githubusercontent.com/ltdrdata/ComfyUI-Manager/main/model-list.json
[ComfyUI-Manager] default cache updated: https://raw.githubusercontent.com/ltdrdata/ComfyUI-Manager/main/custom-node-list.json
[ComfyUI-Manager] default cache updated: https://raw.githubusercontent.com/ltdrdata/ComfyUI-Manager/main/extension-node-map.json
[ComfyUI-Manager] default cache updated: https://raw.githubusercontent.com/ltdrdata/ComfyUI-Manager/main/github-stats.json
FETCH DATA from: F:\StableDiffusion\ComfyUI-Zluda\custom_nodes\ComfyUI-Manager\extension-node-map.json [DONE]
got prompt
model weight dtype torch.float16, manual cast: None
model_type EPS
Using split attention in VAE
Using split attention in VAE
Requested to load SDXLClipModel
Loading 1 new model
loaded completely 0.0 1560.802734375 True
Requested to load SDXL
Loading 1 new model
loaded completely 0.0 4897.0483474731445 True
100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████| 20/20 [00:02<00:00,  8.75it/s]
Requested to load AutoencoderKL
Loading 1 new model
loaded completely 0.0 159.55708122253418 True
Prompt executed in 8.83 seconds

Stopped server
^CTerminate batch job (Y/N)? y

(venv) F:\StableDiffusion\ComfyUI-Zluda>comfyui.bat
*** Checking and updating to new version if possible
Already up to date.
**<------------- And python pop-up came here again** 
Press any key to continue . . .

**Comment by patientx** [comment]
first of all, it is not about the missing package, it is optional. you can install it manually as I said above or completely remove that node pack, (impact pack, just delete the folder under custom_nodes for the moment) secondly why are you trying to close the app via key strokes ? just close the command line window via the X button on the window. No need to ctrl-c or x it.  And try re-running the batch file from the directory again.

 If you want more detail ::

open cmd
cd /d F:\StableDiffusion\ComfyUI-Zluda
call F:\StableDiffusion\ComfyUI-Zluda\venv\Scripts\activate 
.\zluda\zluda.exe -- python main.py --auto-launch --use-quad-cross-attention --verbose

(last one prints extra detail)

**Comment by ReiskaC** [comment]
God, this is frustrating... I've rebuilt everything several times from scratch and even installed ComfyUI proper with ROCm 6.2 nightly under WSL2 / Debian 22.04 as a comparison build. 

Sadly, when Zluda doesn't bork out and works it's abt. 25-40% faster (148s vs. 200-400s for identical workflows run on both) in complex workflows than ROCm WSL-Comfy but at least with a few days of testing I haven't managed to break the WSL-version vs. Zluda borked yet again with the same error. 

Tried all the tricks suggested in the video regarding the DLL reregistering, etc. with no deterministic outcome. Sometimes it works just with a reboot, some times it is enough to close the command prompt window I've been running comfyui.bat from and open a new CMD, sometimes it just doesn't revive itself and requires a full wipe and re-install from GIT. 

I'm not sure, but sometimes it seems its the Zluda-wrapper that may be the cause of this failing as I've been able to start ComfyUI with just python without the Zluda wrapper even after the wrapped restart crashes constantly (ofc. Comfy won't run as the CUDA device is not found, but python doesn't crash when invoked without Zluda-wrapper.
